File: tools/noabbr.py
#!/usr/bin/env python3
"""No abbreviations: nothing can define one, so nothing asks whether one applies.

Usage:
    python3 tools/noabbr.py <file>

tools/retire.py points the twelve rows -- :abbreviate, :noreabbrev,
:unabbreviate, :abclear and their `i` and `c` forms -- at ex_ni, and with them
goes the only way an abbreviation is ever defined.  What is left is the
question: insert mode asks `echeck_abbr()` on ESC, CTRL-O, CTRL-L, Tab, Enter
and every non-word character, and the command line asks `ccheck_abbr()` twice.
With no abbreviation there to find, each asks it for ever and is told no.  So
the questions go, and the sweep takes check_abbr() -- 195 lines -- with the two
wrappers and the handlers.

Each cut folds a test whose answer is now known:

  * `if (echeck_abbr(...)) { ... }`  and  `if (ccheck_abbr(...)) { ... }`
    guard what happens when an abbreviation fired, so the block goes.
  * `!echeck_abbr(x) && c != Ctrl_RSB` is `c != Ctrl_RSB`.
  * `(ccheck_abbr(x) || c == Ctrl_RSB)` is `c == Ctrl_RSB` -- CTRL-] on the
    command line still triggers "an abbreviation", which is to say nothing, and
    still does not insert itself.

What stays: the mapping code's `abbr` parameters and list, which mappings share.
Nothing can put an entry on the abbreviation list any more, and nothing here
pretends that makes the shared code smaller.
"""

import re
import sys
from pathlib import Path

sys.path.insert(0, __file__.rsplit('/', 1)[0])
import cutil

FOLDS = [
    ('insert mode: ESC expanding an abbreviation first', r'^[ \t]*if \(echeck_abbr\(ESC \+ ABBR_OFF\)\)$'),
    ('insert mode: CTRL-O expanding an abbreviation first', r'^[ \t]*if \(echeck_abbr\(Ctrl_O \+ ABBR_OFF\)\)$'),
    ("insert mode: CTRL-L under 'insertmode'", r'^[ \t]*if \(echeck_abbr\(Ctrl_L \+ ABBR_OFF\)\)$'),
    ('insert mode: Tab expanding an abbreviation first', r'^[ \t]*if \(echeck_abbr\(TAB \+ ABBR_OFF\)\)$'),
    ('insert mode: Enter expanding an abbreviation first', r'^[ \t]*if \(echeck_abbr\(c \+ ABBR_OFF\)\)$'),
    ('the command line: a special key expanding an abbreviation', r'^[ \t]*if \(ccheck_abbr\(c \+ ABBR_OFF\)\)$'),
]

LITERAL = [
    ('insert mode: a non-word character inserting unless an abbreviation took it',
     'if (vim_iswordc(c) || (!echeck_abbr((has_mbyte && c >= 0x100) ? (c + ABBR_OFF) : c) && c != Ctrl_RSB))',
     'if (vim_iswordc(c) || c != Ctrl_RSB)'),
    ('the command line: a non-word character expanding an abbreviation',
     '(ccheck_abbr((has_mbyte && c >= 0x100) ? (c + ABBR_OFF) : c) || c == Ctrl_RSB)',
     'c == Ctrl_RSB'),
]


def main():
    if len(sys.argv) != 2:
        sys.exit(__doc__)
    path = Path(sys.argv[1])
    text = path.read_text(errors='surrogateescape')

    for what, pattern in FOLDS:
        try:
            text = cutil.fold_never(text, pattern, 1, re.M)
        except ValueError as e:
            sys.exit('noabbr: %s -- %s' % (what, e))
        print('  noabbr       %s' % what)
    for what, old, new in LITERAL:
        n = text.count(old)
        if n != 1:
            sys.exit('noabbr: %s -- occurs %d times, not once' % (what, n))
        text = text.replace(old, new)
        print('  noabbr       %s' % what)

    # Nothing may still ASK.  The wrappers must have no caller at all, and
    # check_abbr() no caller outside the two wrappers -- which the sweep takes,
    # so a call left inside one of them is not a caller.
    spans = [cutil.find_definition(text, n) for n in ('echeck_abbr', 'ccheck_abbr', 'check_abbr')]
    inside = lambda pos: any(s and s[0] <= pos < s[1] for s in spans)
    calls = [m for m in re.finditer(r'\b[ec]?check_abbr\(', text)
             if not inside(m.start())
             and not re.match(r'static\s+int\s+[ec]?check_abbr\(', text[text.rfind('\n', 0, m.start()) + 1:])]
    if calls:
        sys.exit('noabbr: still asked at:\n    ' + '\n    '.join(
            text[text.rfind('\n', 0, m.start()) + 1:text.find('\n', m.start())].strip()[:100] for m in calls))

    path.write_text(text, errors='surrogateescape')
    print('  noabbr       nothing asks whether an abbreviation applies')


if __name__ == '__main__':
    main()

File: tools/go/internal/cut/noabbr.go
package cut

import (
	"bytes"
	"fmt"
	"io"
	"regexp"
	"strings"

	"slimvim.local/tools/internal/cutil"
)

var abbrFolds = []struct{ what, pattern string }{
	{"insert mode: ESC expanding an abbreviation first",
		`(?m)^[ \t]*if \(echeck_abbr\(ESC \+ ABBR_OFF\)\)$`},
	{"insert mode: CTRL-O expanding an abbreviation first",
		`(?m)^[ \t]*if \(echeck_abbr\(Ctrl_O \+ ABBR_OFF\)\)$`},
	{"insert mode: CTRL-L under 'insertmode'",
		`(?m)^[ \t]*if \(echeck_abbr\(Ctrl_L \+ ABBR_OFF\)\)$`},
	{"insert mode: Tab expanding an abbreviation first",
		`(?m)^[ \t]*if \(echeck_abbr\(TAB \+ ABBR_OFF\)\)$`},
	{"insert mode: Enter expanding an abbreviation first",
		`(?m)^[ \t]*if \(echeck_abbr\(c \+ ABBR_OFF\)\)$`},
	{"the command line: a special key expanding an abbreviation",
		`(?m)^[ \t]*if \(ccheck_abbr\(c \+ ABBR_OFF\)\)$`},
}

var abbrLiteral = []struct{ what, old, new string }{
	{"insert mode: a non-word character inserting unless an abbreviation took it",
		"if (vim_iswordc(c) || (!echeck_abbr((has_mbyte && c >= 0x100) ? (c + ABBR_OFF) : c) && c != Ctrl_RSB))",
		"if (vim_iswordc(c) || c != Ctrl_RSB)"},
	{"the command line: a non-word character expanding an abbreviation",
		"(ccheck_abbr((has_mbyte && c >= 0x100) ? (c + ABBR_OFF) : c) || c == Ctrl_RSB)",
		"c == Ctrl_RSB"},
}

var (
	abbrCall  = regexp.MustCompile(`\b[ec]?check_abbr\(`)
	abbrProto = regexp.MustCompile(`^static\s+int\s+[ec]?check_abbr\(`)
)

// NoAbbr takes away every place an abbreviation could still be expanded.
//
// The last act is the one that matters: nothing may still ASK.  The two
// wrappers must have no caller at all and check_abbr() none outside them --
// which the sweep takes, so a call left INSIDE one of them is not a caller,
// and that is why the scan excludes the three definitions' own spans rather
// than counting mentions.
func NoAbbr(text []byte, w io.Writer) ([]byte, error) {
	for _, f := range abbrFolds {
		out, err := cutil.FoldNever(text, f.pattern, 1)
		if err != nil {
			return nil, fmt.Errorf("noabbr: %s -- %v", f.what, err)
		}
		text = out
		fmt.Fprintf(w, "  noabbr       %s\n", f.what)
	}
	for _, l := range abbrLiteral {
		n := bytes.Count(text, []byte(l.old))
		if n != 1 {
			return nil, fmt.Errorf("noabbr: %s -- occurs %d times, not once", l.what, n)
		}
		text = bytes.ReplaceAll(text, []byte(l.old), []byte(l.new))
		fmt.Fprintf(w, "  noabbr       %s\n", l.what)
	}

	blanked := cutil.Blank(text)
	type span struct{ a, z int }
	var spans []span
	for _, n := range []string{"echeck_abbr", "ccheck_abbr", "check_abbr"} {
		if a, z, ok := cutil.FindDefinition(text, blanked, n); ok {
			spans = append(spans, span{a, z})
		}
	}
	inside := func(pos int) bool {
		for _, s := range spans {
			if s.a <= pos && pos < s.z {
				return true
			}
		}
		return false
	}
	lineAt := func(pos int) []byte {
		a := bytes.LastIndexByte(text[:pos], '\n') + 1
		z := bytes.IndexByte(text[pos:], '\n')
		if z < 0 {
			return text[a:]
		}
		return text[a : pos+z]
	}
	var still []string
	for _, m := range abbrCall.FindAllIndex(text, -1) {
		if inside(m[0]) {
			continue
		}
		a := bytes.LastIndexByte(text[:m[0]], '\n') + 1
		if abbrProto.Match(text[a:]) {
			continue
		}
		line := strings.TrimSpace(string(lineAt(m[0])))
		if len(line) > 100 {
			line = line[:100]
		}
		still = append(still, line)
	}
	if len(still) > 0 {
		return nil, fmt.Errorf("noabbr: still asked at:\n    %s", strings.Join(still, "\n    "))
	}

	fmt.Fprintln(w, "  noabbr       nothing asks whether an abbreviation applies")
	return text, nil
}
